### PR TITLE
add new predicate: max replicas limit per node

### LINKED ISSUE #71930: add a new predicate: max replicas limit per node
**What would you like to be added**:
add a new predicate: max replicas limit per node 

**Why is this needed**:
I want to add a new predicate that supports this scene:
when we deploy a statefulset, we need control the replicas number of statefulset. Considered that the network, memory, disk IO are limited. we could appoint the number of replicas to deploy in one node. Anti-affinity could not help us control the accurate number clearly.

/kind feature

### PR DESCRIPTION
**Pull request by zhangxiaoyu-zidif**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake
/kind feature

**What this PR does / why we need it**:
if we want to deploy an app on the cluster, we may want set a max replicas number per node of this app, so that on that node only limited number replicas can be deployed. 

**Which issue(s) this PR fixes**:

Fixes #71930 
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
add a new scheduling predicate: MaxReplicasPerHost
```

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -680,6 +686,102 @@ func (c *VolumeZoneChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
...
+	return c.predicate
+}
+
+func (c *ReplicasSpreadChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	// check the limited number of replicas
+	if pod.Annotations != nil {
+		if maxReplicasNumber, ok := pod.Annotations[MaxReplicasPerHost]; ok {
+			maxReplicasNumberValue, err := strconv.Atoi(maxReplicasNumber)
+			if err != nil {
+				klog.Errorf("failed to parse MaxReplicasPerHost [%s] to a number: %v", maxReplicasNumber, err)
```
Comment (k82cn) [inline pkg/scheduler/algorithm/predicates/predicates.go]: if err , does `maxReplicasNumberValue ` valid to do other action?

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -680,6 +686,102 @@ func (c *VolumeZoneChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
...
+				count++
+			}
+		}
+	}
+	return count
+}
+
+// ownerReferencesEqual checks  if the two slices have the same elements.
+func ownerReferencesEqual(list1 []metav1.OwnerReference, list2 []metav1.OwnerReference) bool {
+	m := map[metav1.OwnerReference]int{}
```
Comment (k82cn) [inline pkg/scheduler/algorithm/predicates/predicates.go]: should we use `controlled` owner reference? how about deployment/rs cases?

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 3]
```diff
@@ -680,6 +686,102 @@ func (c *VolumeZoneChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
...
+
+func (c *ReplicasSpreadChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	// check the limited number of replicas
+	if pod.Annotations != nil {
+		if maxReplicasNumber, ok := pod.Annotations[MaxReplicasPerHost]; ok {
+			maxReplicasNumberValue, err := strconv.Atoi(maxReplicasNumber)
+			if err != nil {
+				klog.Errorf("failed to parse MaxReplicasPerHost [%s] to a number: %v", maxReplicasNumber, err)
+			}
+			podsOnNode := nodeInfo.Pods()
```
Comment (k82cn) [inline pkg/scheduler/algorithm/predicates/predicates.go]: should we also support zone?

**Comment by wgliang** [comment]
Thanks @zhangxiaoyu-zidif 
Because the current scheduling algorithm is pod by pod, I would like to know how the difference between setting max replicas and using Pod's affinity? In fact, this is more like a special need for a real scene.

**Review by yastij — CHANGES_REQUESTED** [review]
I'm not sure we want to introduce such change, this could be done with hard anti-affinity. Do have real use cases where you'd want max replicas per node and not just use anti-affinity ? 

looping @bsalamat

**Comment by zhangxiaoyu-zidif** [comment]
@wgliang  @yastij  @k82cn  thanks all, let me open an issue to discuss this.

**Review by resouer — CHANGES_REQUESTED** [review]
This use case may be valid, but I would suggest to create an issue firstly so folks can discuss all scenarios and potential workarounds. :-)

Also, we may want to avoid adding specific annotation in Pod unless it's a well-defined alpha feature with roll out plan.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 3 above)
Comment (zhangxiaoyu-zidif) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Good point, thanks. although I have not consider it here, but it is worthing considering.

**Comment by zhangxiaoyu-zidif** [comment]
> od by pod, I would like to know how the difference between setting max replicas and using Pod's affinity? In fact, this is more like a special need for a real scene.

in fact, we could not use anti-affinity to make the replicas number accurate. in large scale cluster, with the same configurations, we need to make sure that the replicas number of one application per node. Disk IO, network bandwidth and so on are all pre-planed.  Anti-affinity is a indistinct feature.

**Pull request closed without merging**